src/Formats/CapnProtoSerializer.cpp

```diff
@@ -0,0 +1,1479 @@
+            }
+            else
+            {
+                auto all_values = enum_values.getValues();
+                if (all_values.size() != enumerants.size())
+                    throw Exception(
+                        ErrorCodes::CAPN_PROTO_BAD_CAST,
+                        "The set of names in Enum from CapnProto schema is different from the set of names in ClickHouse Enum");
+
+                std::unordered_map<String, EnumType> ch_name_to_value;
+                for (auto & [name, value] : all_values)
+                    ch_name_to_value[to_lower ? boost::algorithm::to_lower_copy(name) : name] = value;
+
+                for (auto enumerant : enumerants)
+                {
+                    String capnp_name = enumerant.getProto().getName();
+                    UInt16 capnp_value = enumerant.getOrdinal();
+                    auto it = ch_name_to_value.find(to_lower ? boost::algorithm::to_lower_copy(capnp_name) : capnp_name);
+                    if (it == ch_name_to_value.end())
+                        throw Exception(
+                            ErrorCodes::CAPN_PROTO_BAD_CAST,
+                            "The set of names in Enum from CapnProto schema is different from the set of names in ClickHouse Enum");
+
+                    ch_to_capnp_values[it->second] = capnp_value;
+                    capnp_to_ch_values[capnp_value] = it->second;
+                }
+            }
+        }
+
+        void writeRow(const ColumnPtr & column, std::unique_ptr<FieldBuilder> &, capnp::DynamicStruct::Builder & parent_struct_builder, UInt32 slot_offset, size_t row_num) override
+        {
+            parent_struct_builder.getBuilderImpl().setDataField<UInt16>(slot_offset, getValue(column, row_num));
+        }
+
+        void writeRow(const ColumnPtr & column, std::unique_ptr<FieldBuilder> &, capnp::DynamicList::Builder & parent_list_builder, UInt32 array_index, size_t row_num) override
+        {
+            parent_list_builder.getBuilderImpl().setDataElement<UInt16>(array_index, getValue(column, row_num));
+        }
+
+        void readRow(IColumn & column, const capnp::DynamicStruct::Reader & parent_struct_reader, UInt32 slot_offset) override
+        {
+            insertValue(column, parent_struct_reader.getReaderImpl().getDataField<UInt16>(slot_offset));
+        }
+
+        void readRow(IColumn & column, const capnp::DynamicList::Reader & parent_list_reader, UInt32 array_index) override
+        {
+            insertValue(column, parent_list_reader.getReaderImpl().getDataElement<UInt16>(array_index));
+        }
+
+    private:
+        UInt16 getValue(const ColumnPtr & column, size_t row_num)
+        {
+            EnumType enum_value = assert_cast<const ColumnVector<EnumType> &>(*column).getElement(row_num);
+            if (enum_comparing_mode == FormatSettings::CapnProtoEnumComparingMode::BY_VALUES)
+                return static_cast<UInt16>(enum_value);
+            return ch_to_capnp_values[enum_value];
```

Comment:
But this map `ch_to_capnp_values ` should contain all clickhouse enum values. In `CapnProtoEnumSerializer` constructor we check that Capnp enum and CH enum have the same names and for each name we map CH value to Capnp value. Should not be a problem here 

> E.g. I just got that to happen with ALTER MODIFY COLUMN that changes type to a smaller enum.

I guess this will just procude incorrect data, you will not be able to read from this table: https://pastila.nl/?051c29b6/acccb123e943e9fb0b5b752ac44e8896

So, I don't think we should think about this case

Reply:
> you will not be able to read from this table: https://pastila.nl/?051c29b6/acccb123e943e9fb0b5b752ac44e8896

This exception happens at formatting time, I think. There's no exception if you cast invalid enum to an int: https://pastila.nl/?015c674c/bd77965f77b9e868b0b2b197a60b8443 . So if we don't add a check here, such invalid values will silently turn into zeroes. Seems worse than exception.